proposals/companion-vals.md

```diff
@@ -0,0 +1,198 @@
+# Companion values
+
+* **Type**: Design proposal
+* **Author**: Roman Sakno
+* **Contributors**: Roman Sakno
+* **Status**: Under consideration
+* **Prototype**: Not started
+
+## Synopsis
+Composition and aggregation are frequently used patterns in OOP that helps to create more complex logic using combination of other projects with respect to encapsulation. Kotlin offers several ways to simplify aggregation in some cases:
+
+* [Implementation by Delegation](https://kotlinlang.org/docs/reference/delegation.html#implementation-by-delegation) under the hood aggregates original implementation of interface and delegates method calls of top-level class to aggregated implementation
+* [Companion object](https://kotlinlang.org/docs/reference/object-declarations.html#companion-objects) allows to aggregate singleton object at class-level (not instance level) and make their members to be accessible without instantiation of outer class.
+
+But there is no way to achieve the same convenience as provided by **companion** objects at instance-level rather than class-level.
+
+The main idea of this proposal is to give implicit access to members of aggregated/composed object rather than existing **val**/**var** declarations with dot notation for accessing their members.
+
+## Language implementation
+**companion** keyword can be applied to **val** or **var** declarations inside of class declaration or at constructor parameter level. This keyword indicates that all members of such value will be accessible implicitly inside of class.
+
+```kotlin
+class Derived(a: Int, b: Int): Base2(a), Serializable{
+  private companion val base1 = Base1(a + b) //explicitly defined name
+  
+  fun foo() = baz() //baz() from Base1
+  fun bar() = base1.baz() //the same but with explicit receiver
+}
+```
+or shorter,
+```kotlin
+class Derived(a: Int, b: Int): Base2(a), Serializable{
+  private companion val: Base1(a + b) //like companion object without explicit name
+  
+  fun foo() = baz() //baz() from Base1
+  fun bar() = this@Base1.baz() //the same but with explicit receiver
+}
+```
+Class may have more than one companion value.
+
+### Constructor parameter
+Ability to aggregate object passed as constructor parameter:
+```kotlin
+class Derived(private companion val b: Base1): Base2(a), Serializable
```

Comment:
Actually looks somehow related to typeclasses proposal that also tries to provide ad-hoc polymorphism.

Reply:
Not sure about typeclasses. This declaration has the same behavior as declaration of companion value at class level.
```kotlin
class Derived(private companion val b: Base1){ }
```
is equal to
```kotlin
class Derived(b: Base1){
  private companion val: Base1 = b
}
```